pubsub/gcloud/aio/pubsub/metrics_agent.py

```diff
@@ -12,3 +15,30 @@ def increment(self,
                   metric: str,
                   value: float = 1) -> None:
         pass
+
+
+class PrometheusMetrics:
+    def __init__(self, *, namespace: str = '',
+                 subsystem: str = '') -> None:
+        self.consume = prometheus_client.Counter(
+            'pubsub_consume',
+            'Counter of the outcomes of PubSub message consume attempts',
+            ['outcome'], namespace=namespace, subsystem=subsystem)
+
+        self.consume_latency = prometheus_client.Histogram(
+            'pubsub_consume_latency',
+            'Histogram of PubSub message consume latencies', ['aspect'],
+            namespace=namespace, subsystem=subsystem)
+
+        self.batch_fail = prometheus_client.Counter(
+            'pubsub_batch_failed',
+            'Counter for failures to process PubSub message batches',
+            ['component'], namespace=namespace, subsystem=subsystem)
+
+        self.messages_processed = prometheus_client.Counter(
+            'pubsub_messages_processed',
+            'Counter of successfully processed messages',
+            ['component'], namespace=namespace, subsystem=subsystem)
+
+
+PROMETHEUS = PrometheusMetrics(namespace='gcloud', subsystem='aio')
```

Comment:
Please double check this gets converted to `subsystem='rest'` at build time.

Reply:
Hmm, should this be `namespace='gcloud_aio', subsystem='pubsub'`? I'm trying to think of whether there would ever be non-pubsub metrics here -- probably not?

Reply:
I was thinking the same thing!

Reply:
FYI `gcloud_rest` is not supported, and this is within an `if rest` block so we're all covered.

Either way the metrics will turn out as `gcloud_aio_pubsub_subscriber` whether `subsystem` is `subscriber` or `pubsub`. I'm inclined to use `subsystem=pubsub` to allow this to be copy-pasted into other clients like `subsystem=storage`. IIRC the `namespace` and `subsystem` construction here is only relevant in code. It is not transmitted as part of the metric to Prometheus.